src/duration.rs

```diff
@@ -498,16 +505,58 @@ impl Duration {
     }
 
     fn parse_time(bytes: &[u8], offset: usize, config: &TimeConfig) -> Result<Self, ParseError> {
-        let t = crate::time::PureTime::parse(bytes, offset, config)?;
+        let byte_len = bytes.len();
+        if byte_len - offset < 5 {
+            return Err(ParseError::TooShort);
+        }
+        const HOUR_NUMERIC_LIMIT: i64 = 24 * 10i64.pow(8);
+        let mut hour: i64 = 0;
+
+        let mut chunks = bytes
+            .get(offset..)
+            .ok_or(ParseError::TooShort)?
+            .splitn(2, |&byte| byte == b':');
+
+        // can just use `.split_once()` in future maybe, if that stabilises
+        let (hour_part, mut remaining) = match (chunks.next(), chunks.next(), chunks.next()) {
+            (_, _, Some(_)) | (None, _, _) => unreachable!("should always be 1 or 2 chunks"),
+            (Some(_hour_part), None, _) => return Err(ParseError::InvalidCharHour),
+            (Some(hour_part), Some(remaining), None) => (hour_part, remaining),
+        };
 
-        if bytes.len() > t.position {
+        for byte in hour_part {
+            let h = *byte - b'0';
+            if h > 9 {
+                return Err(ParseError::InvalidCharHour);
+            }
+            hour = (hour * 10) + (h as i64);
+        }
+        if hour > HOUR_NUMERIC_LIMIT {
+            return Err(ParseError::DurationHourValueTooLarge);
+        }
```

Comment:
Just as a final thought here, I think it's possible for us to overflow if `hour_part` is too many bytes.

I think in a previous version you had a check on the maximum length of `hour_part`, probably we just dropped that during rounds of simplification. Maybe we should reintroduce that?

Reply:
nice catch. I just added that in again.